### PR TITLE
Setup curl handle once and reuse it between API requests.

### PR DESCRIPTION
**Pull request by thijskh**
This saves us on setup time and cURL will reuse the HTTP connection.

Not yet fully tested, do not merge.

### CODE REVIEW

**Comment by meisterT** [comment]
I also tried this once and the difference wasn't that big but I think we should do it anyway. I'll measure with my setup and post numbers here.

**Comment by thijskh** [comment]
This is currently not complete since we can change api credentials between calls. Can fix that if we decide to go through with it.

File: judge/judgedaemon.main.php [hunk 1]
```diff
@@ -46,6 +46,28 @@ function read_credentials()
...
+        curl_setopt($curl_handle, CURLOPT_RETURNTRANSFER, true);
+    }
+}
+
+function close_curl_handle()
+{
+    global $curl_handle;
+    if ( ! empty($curl_handle) ) {
+        curl_close($curl_handle);
+    }
```
Comment (meisterT) [inline judge/judgedaemon.main.php]: currently it's not necessary from the calls that I can see, but we should reset the handle here

File: judge/judgedaemon.main.php [hunk 2]
```diff
@@ -46,6 +46,28 @@ function read_credentials()
     }
 }
 
+
+function setup_curl_handle()
+{
+    global $curl_handle, $restuser, $restpass;
```
Comment (meisterT) [inline judge/judgedaemon.main.php]: one curl_handle is not enough. We need one per endpoint, right?

File: judge/judgedaemon.main.php (hunk 1 above)
Comment (thijskh) [reply judge/judgedaemon.main.php]: agreed

File: judge/judgedaemon.main.php (hunk 2 above)
Comment (thijskh) [reply judge/judgedaemon.main.php]: yes, could create one per endpoint, or move the basic auth credential setting out of this function

File: judge/judgedaemon.main.php [hunk 3]
```diff
@@ -70,39 +92,38 @@ function request(string $url, string $verb = 'GET', string $data = '', bool $fai
...
-    $ch = curl_init($url);
-    curl_setopt($ch, CURLOPT_USERAGENT, "DOMjudge/" . DOMJUDGE_VERSION);
-    curl_setopt($ch, CURLOPT_HTTPAUTH, CURLAUTH_BASIC);
-    curl_setopt($ch, CURLOPT_USERPWD, $restuser . ":" . $restpass);
-    curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
+    curl_setopt($curl_handle, CURLOPT_URL, $url);
+
     if ($verb == 'POST') {
-        curl_setopt($ch, CURLOPT_POST, true);
+        curl_setopt($curl_handle, CURLOPT_POST, true);
```
Comment (meisterT) [inline judge/judgedaemon.main.php]: Adding 
`curl_setopt($curl_handle, CURLOPT_POST, false);`
two lines above gets past the second request. It still ends up with 404 later, but I'll keep digging what else we have to reset.

File: judge/judgedaemon.main.php [hunk 4]
```diff
@@ -70,39 +92,38 @@ function request(string $url, string $verb = 'GET', string $data = '', bool $fai
...
         if (is_array($data)) {
-            curl_setopt($ch, CURLOPT_HTTPHEADER, array('Content-Type: multipart/form-data'));
+            curl_setopt($curl_handle, CURLOPT_HTTPHEADER, ['Content-Type: multipart/form-data']);
         }
     } elseif ($verb == 'PUT' || $verb == 'DELETE') {
-        curl_setopt($ch, CURLOPT_CUSTOMREQUEST, $verb);
+        curl_setopt($curl_handle, CURLOPT_CUSTOMREQUEST, $verb);
```
Comment (meisterT) [inline judge/judgedaemon.main.php]: this is the other option which has to be reset between requests

File: judge/judgedaemon.main.php (hunk 1 above)
Comment (thijskh) [reply judge/judgedaemon.main.php]: Agreed, `curl_close()` promises to "delete the handle" but that's not fully resetting `$curl_handle` apparently.

**Comment by thijskh** [comment]
My dev environment is a bit borked so I made this PR more as a sketch. Have added the things above except for the multi-endpoint support.

**Pull request merged**